src/paillier.rs

```diff
@@ -78,3 +123,116 @@ impl PaillierDecryptionKey {
+    pub(crate) fn get_random_safe_prime_512<R: RngCore + CryptoRng>(rng: &mut R) -> BigNumber {
+        BigNumber::safe_prime_from_rng(PRIME_BITS, rng)
+    }
+
+    // Generate safe primes from a file. Usually, generating safe primes takes
+    // awhile (0-5 minutes per 512-bit safe prime on my laptop, average 50 seconds)
+    #[cfg(test)]
+    lazy_static::lazy_static! {
+        static ref POOL_OF_PRIMES: Vec<BigNumber> = get_safe_primes_from_file();
+    }
+
+    #[cfg(test)]
+    fn get_safe_primes_from_file() -> Vec<BigNumber> {
+        crate::safe_primes_512::SAFE_PRIMES
+            .iter()
+            .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
+            .filter(|prime| prime.bit_length() == PRIME_BITS)
+            .collect()
+    }
+
+    /// Sample a safe prime from a precompiled list. For testing purposes only!!
+    #[cfg(test)]
+    pub(crate) fn get_prime_from_pool_insecure<R: RngCore + CryptoRng>(rng: &mut R) -> BigNumber {
+        POOL_OF_PRIMES[rng.gen_range(0..POOL_OF_PRIMES.len())].clone()
+    }
+
+    /// Sample a pair of independent, non-matching safe primes from a precompiled list.
+    /// For testing purposes only!!
+    #[cfg(test)]
+    pub(crate) fn get_prime_pair_from_pool_insecure<R: RngCore + CryptoRng>(
+        rng: &mut R,
+    ) -> (BigNumber, BigNumber) {
+        let p = get_prime_from_pool_insecure(rng);
+        loop {
+            let q = get_prime_from_pool_insecure(rng);
+            if p != q {
+                break (p, q);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use libpaillier::unknown_order::BigNumber;
+    use rand::{
+        rngs::{OsRng, StdRng},
+        Rng, SeedableRng,
+    };
+
+    use crate::parameters::PRIME_BITS;
+
+    use super::{prime_gen, PaillierDecryptionKey};
+
+    fn rng() -> StdRng {
+        let mut seeder = OsRng;
+        let seed = seeder.gen();
+        eprintln!("seed: {:?}", seed);
+        StdRng::from_seed(seed)
+    }
+
+    #[test]
+    fn get_random_safe_prime_512_produces_safe_primes() {
+        let mut rng = rng();
+        let p = prime_gen::get_random_safe_prime_512(&mut rng);
+        assert!(p.is_prime());
+        let q: BigNumber = (p - 1) / 2;
+        assert!(q.is_prime());
+    }
+
+    #[test]
+    fn paillier_keygen_produces_good_primes() {
+        let mut rng = rng();
+
+        let (decryption_key, p, q) = PaillierDecryptionKey::new(&mut rng).unwrap();
+
+        assert!(p.is_prime());
+        assert!(q.is_prime());
+
+        let safe_p: BigNumber = (&p - 1) / 2;
+        assert!(safe_p.is_prime());
+        let safe_q: BigNumber = (&q - 1) / 2;
+        assert!(safe_q.is_prime());
+
+        assert_eq!(p.bit_length(), PRIME_BITS);
+        assert_eq!(q.bit_length(), PRIME_BITS);
+
+        let modulus = &p * &q;
+        assert_eq!(decryption_key.0.n(), &modulus);
+        assert_eq!(modulus.bit_length(), 2 * PRIME_BITS);
+    }
+
+    #[test]
+    #[ignore = "slow"]
```

Comment:
note: This never gets run right now, but once we address #65 it should be run sometimes.